modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/RunningQueriesIntegrationTest.java

```diff
@@ -234,4 +254,95 @@ public void testCancelByRemoteFragment() throws IgniteCheckedException {
 
         GridTestUtils.assertThrowsAnyCause(log, () -> fut.get(100), IgniteSQLException.class, "The query was cancelled while executing.");
     }
+
+    /** */
+    @Test
+    public void testLongPlanningTimeout() throws Exception {
+        Stream.of("T1", "T2").forEach(tblName -> {
+            sql(String.format("CREATE TABLE %s(A INT, B INT)", tblName));
+
+            IgniteTable tbl = (IgniteTable)queryProcessor(client).schemaHolder().schema("PUBLIC").getTable(tblName);
+            IgniteIndex oldPk = tbl.getIndex("_key_PK");
+            tbl.addIndex(new SlowCollationIndex(oldPk));
+
+            sql(String.format("INSERT INTO %s(A, B) VALUES (1, 1)", tblName));
+        });
+
+        String longJoinQry = "SELECT * FROM T1 JOIN T2 ON T1.A = T2.A";
+
+        try {
+            AtomicReference<List<List<?>>> res = new AtomicReference<>();
+            GridTestUtils.assertTimeout(3 * PLANNER_TIMEOUT, TimeUnit.MILLISECONDS, () -> {
+                res.set(sql(longJoinQry));
+            });
+
+            assertNotNull(res.get());
+            assertFalse(res.get().isEmpty());
+        }
+        catch (Exception e) {
+            assertTrue("Unexpected exception: " + e, e instanceof RelOptPlanner.CannotPlanException);
+        }
+    }
+
+    /** */
+    private static class SlowCollationIndex implements IgniteIndex {
```

Comment:
Perhaps it's worth to create `DelegatingIgniteIndex` class in `AbstractBasicIntegrationTest` and reuse it in this test and `IndexScanlIntegrationTest` to remove boilerplate code. WDYT?